### PR TITLE
Add new `HorizontalTab` component to support sublabels, icons and disabled attribute

### PR DESCRIPTION
**Pull request by mayank99**
This is a difficult one to do without breaking changes. The current implementation has a `labels` prop which accepts a `ReactNodeArray`, and places each of those labels inside a `button` (which already has classname, event handler, etc applied).

Here's what I've tried:

<details>
<summary> ❌ Approach 1: <code>TabContent</code> component (fragment), passed into <code>labels</code> array.</summary>

Initially I wanted to make a `Tab` component, but because it needs to be backwards compatible, I was not able to make the `button` element part of it. It would work ok for the sublabel and icons part, but the `disabled` attribute needs to be applied on `button`, so I got stuck there. See [`be60186` (#181)](https://github.com/iTwin/iTwinUI-react/pull/181/commits/be6018600234e709d6deb51c52db00f0bfe5885a).

Maybe I could look into type guards? Not sure.

</details>

<details>
<summary> ❌ Approach 2: <code>HorizontalTab</code> object, passed into <code>labels</code> array.</summary>

I thought, maybe the `labels` array could accept both ReactNodes and an object like this:
```tsx
<HorizontalTabs
   labels=[
     { label: 'Label with icon', icon: <SvgPlaceholder /> },
     { label: 'Label with description', sublabel: 'Description' },
     { label: 'Label with icon and disabled', icon: <SvgPlaceholder />, disabled: true },
   ]
```

It allows me to easily access the properties. But the whole approach seems hacky, and the code is not very nice. Also ran into some TS issues. See [`7506527` (#181)](https://github.com/iTwin/iTwinUI-react/pull/181/commits/750652762a2338e734ebf9b26c77b1c2b6b25195).

</details>

<details>
<summary> ❌ Approach 3: <code>HorizontalTab</code> component, passed into new <code>tabs</code> prop.</summary>

This will preserve the old `labels` functionality (marked as deprecated), and allow for `button` to be part of the Tab component.

I used union types and `in` keyword to _require_ either `labels` or `tabs` props but not both at the same time.

See [`128be1e` (#181)](https://github.com/iTwin/iTwinUI-react/pull/181/commits/128be1eed20b6cfb535955d4d7086a0ce7055a83)

</details>

<details>
<summary><strong> ✔ Approach 4: <code>HorizontalTab</code> component, passed into <code>labels</code> array.</strong></summary>

This is a _slight_ breaking change but it preserves the old API with the caveat that strings will be automatically wrapped with the new `HorizontalTab` component, and for anything else (e.g. icons or sublabels), the user is expected to manually pass `HorizontalTab`s in the `labels` array.

See [`8f7c0fc` (#181)](https://github.com/iTwin/iTwinUI-react/pull/181/commits/8f7c0fcc7449a9a09adf3f65e0a3d565d8dff475)

</details>

image

---

<details>
<summary><strong>Checklist</strong></summary>

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [x] Add an entry to the changelog for any new components and changes
- [x] Add screenshots of the key elements of the component
- [x] Add any additional comments describing the features you've implemented
</details>

### CODE REVIEW

**Comment by bentleyvk** [comment]
Would it be a breaking change if we checked whether label is a string or JSX? If it is a string then we would wrap it with our Tab component if it is not then just add required props using cloneElement?
Or it will be a breaking change because some one might have passed e.g. icon + text...

**Comment by mayank99** [comment]
> Would it be a breaking change if we checked whether label is a string or JSX? If it is a string then we would wrap it with our Tab component if it is not then just add required props using cloneElement?

I've decided to go with exactly that. The new `HorizontalTab` component is expected to be passed into the `labels` array, preserving the old API.

> Or it will be a breaking change because some one might have passed e.g. icon + text...

Yup, it will be a visual breaking change if someone passes an icon without using `HorizontalTab`. This is what it looks like:
image

File: src/core/HorizontalTabs/HorizontalTabs.tsx [hunk 1]
```diff
@@ -140,16 +158,28 @@ export const HorizontalTabs = (props: HorizontalTabsProps) => {
...
-              >
-                {label}
-              </button>
+              {typeof label === 'string' ? (
+                <HorizontalTab
+                  label={label}
+                  className={cx('iui-tab', {
+                    'iui-active': index === currentIndex,
+                  })}
+                  onClick={onClick}
```
Comment (gretanausedaite) [inline src/core/HorizontalTabs/HorizontalTabs.tsx:167]: I think active tab should be changed on click. 
horizontal-tabs

Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTabs.tsx:167]: It works on my side 🤔 Maybe you need to run `yarn install` for the latest css.

File: src/core/HorizontalTabs/HorizontalTab.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  label?: React.ReactNode;
+  /**
+   * Secondary label shown below the main label.
+   * Only shown if `size` of `HorizontalTabs` is set to large.
+   */
+  sublabel?: React.ReactNode;
+  /**
+   * Svg icon shown before the labels.
+   */
+  icon?: JSX.Element;
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTab.tsx:24]: Maybe lets do `startIcon`, so its obviuos where its put (and maybe we will need endIco in future :DD)

File: src/core/HorizontalTabs/HorizontalTab.test.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  expect(label.firstElementChild?.textContent).toBe('Tab label');
+});
+
+it('should render with sublabel', () => {
+  const { container } = render(
+    <HorizontalTab label='Tab label' sublabel='Sub-label' />,
+  );
+  expect(container.querySelector('button.iui-tab')).toBeTruthy();
+
+  const label = container.querySelector('.iui-tab-label') as HTMLElement;
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTab.test.tsx:26]: maybe check the `> div` selector in here, to be specific

File: src/core/HorizontalTabs/HorizontalTab.test.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import { HorizontalTab } from './HorizontalTab';
+
+it('should render in its most basic state', () => {
+  const { container } = render(<HorizontalTab label='Tab label' />);
+  expect(container.querySelector('button.iui-tab')).toBeTruthy();
+
+  const label = container.querySelector('.iui-tab-label') as HTMLElement;
+  expect(label).toBeTruthy();
+  expect(label.firstElementChild?.textContent).toBe('Tab label');
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTab.test.tsx:17]: just `textContent` on the label was not working?

File: src/core/HorizontalTabs/HorizontalTab.test.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  const { container } = render(<HorizontalTab label='Tab label' disabled />);
+
+  const tab = container.querySelector('button.iui-tab') as HTMLButtonElement;
+  expect(tab).toBeTruthy();
+  expect(tab.disabled).toBeTruthy();
+
+  const label = container.querySelector('.iui-tab-label') as HTMLElement;
+  expect(label).toBeTruthy();
+  expect(label.firstElementChild?.textContent).toBe('Tab label');
+});
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTab.test.tsx:60]: Test for children would be useful

File: src/core/HorizontalTabs/HorizontalTabs.tsx [hunk 6]
```diff
@@ -32,27 +34,48 @@ export type HorizontalTabsProps = {
    * @default 'default'
    */
   type?: 'default' | 'borderless' | 'pill';
+  /**
+   * Modify height of tabs. Use 'large' when tabs have sublabels.
+   * @default 'default'
+   */
+  size?: 'default' | 'large';
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTabs.tsx:41]: ahh.. This one..would be great if it works same as we have done with Menu..

File: src/core/HorizontalTabs/HorizontalTabs.tsx [hunk 7]
```diff
@@ -140,16 +165,28 @@ export const HorizontalTabs = (props: HorizontalTabsProps) => {
...
-                {label}
-              </button>
+              {typeof label === 'string' ? (
+                <HorizontalTab
+                  label={label}
+                  className={cx({
+                    'iui-active': index === currentIndex,
+                  })}
+                  onClick={onClick}
+                  role='tab'
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTabs.tsx:175]: role not needed anymore

File: src/core/HorizontalTabs/HorizontalTab.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ *   <HorizontalTab label='Label 2' icon={<SvgPlaceholder />} />,
+ * ];
+ */
+export const HorizontalTab = (props: HorizontalTabProps) => {
+  const { label, sublabel, icon, children, className, ...rest } = props;
+
+  useTheme();
+
+  return (
+    <button className={cx('iui-tab', className)} role='tab' {...rest}>
```
Comment (veekeys) [inline src/core/HorizontalTabs/HorizontalTab.tsx:50]: What I noticed, now keyboard navigation works (just not the way, we want, I guess). Pressin `Tab` navigates through the tabs (cause of buttons)

File: src/core/HorizontalTabs/HorizontalTabs.tsx (hunk 6 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTabs.tsx:41]: cant. Size is controlled in tabs, sublabel is controlled in individual tab

File: src/core/HorizontalTabs/HorizontalTab.tsx (hunk 8 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.tsx:50]: > now

It's the same as current behavior: https://itwin.github.io/iTwinUI-react/?path=/story/core-horizontaltabs

Comment (veekeys) [reply src/core/HorizontalTabs/HorizontalTab.tsx:50]: default tabs story is stupid, when it shows same content...
but ok, we can leave for this PR and change with the next, I guess..

File: src/core/HorizontalTabs/HorizontalTabs.tsx (hunk 6 above)
Comment (veekeys) [reply src/core/HorizontalTabs/HorizontalTabs.tsx:41]: I know, thats what I am concerned about now..

File: src/core/HorizontalTabs/HorizontalTab.tsx (hunk 2 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.tsx]: Done

File: src/core/HorizontalTabs/HorizontalTab.test.tsx (hunk 3 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.test.tsx:26]: Done

File: src/core/HorizontalTabs/HorizontalTab.test.tsx (hunk 4 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.test.tsx:17]: I thought it needed the exact element, but it works 😮

File: src/core/HorizontalTabs/HorizontalTab.test.tsx (hunk 5 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.test.tsx:70]: Added

File: src/core/HorizontalTabs/HorizontalTab.test.tsx (hunk 4 above)
Comment (veekeys) [reply src/core/HorizontalTabs/HorizontalTab.test.tsx:17]: textContent can go deeper..

File: src/core/HorizontalTabs/HorizontalTab.tsx (hunk 8 above)
Comment (mayank99) [reply src/core/HorizontalTabs/HorizontalTab.tsx:50]: Oh it's that weird storybook bug where state doesn't work on the first story you visit.

Try going to https://itwin.github.io/iTwinUI-react/ and then to the default tabs story

Comment (veekeys) [reply src/core/HorizontalTabs/HorizontalTab.tsx:50]: what a.........

**Review by veekeys — APPROVED** [review]

**Pull request merged**